### ISSUE TITLE
:terminal bold bright inconsistency in some terminals (iterm2)

### ISSUE THREAD

**Issue by bluz71** [neovim/neovim#11335]
Hello Neovim and @bfredl,

The recent merging of #10994 is very nice. It fixes many of the issues I reported in #7018. So thanks indeed. It is a very big improvement.

However, there is a quirk I notice. Bolded named colors are being displayed with non-bolded colors.

Steps to reproduce issue:

- Use latest nightly
- Enable `termguicolors`
- Launch `:terminal`
- Type `git log --format='%C(bold green)%h%C(bold red)%d%C(reset)' `

This should be using color `10` (bold green) and color `9` (bold), but instead it is displaying with color `2` (green) and `1` (red).

Compare and contrast that `git log` command outside of nvim terminal, correct bolded colors will be used.

Doing `git log --format='%C(10)%h%C(9)%d%C(reset)' ` does work, however quite a few utility which use **bolded named colors** are broken.

Many thanks.

**Comment by bfredl** [comment]
Hmm, nvim is supposed to just forward whatever info the child program is setting. I e if it is sending color 2+bold it would forward that to the host terminal, which can interpret that as color 10 if it wants to.

**Comment by bfredl** [comment]
I'm getting the expected behavior (bold bright with and without nvim) in an xterm (and pangoterm, which uses TERM=xterm).

Always mention your terminal emulator and TERM value in a bug report, otherwise debugging the issue is impossible (it is in the template for a reason)

**Comment by bluz71** [comment]
@bfredl,

Sorry, I left off that very important detail. My bad.

The failure happens with [iTerm2](https://iterm2.com) on Mac.

In the meantime I also tested on [GNOME Terminal](https://help.gnome.org/users/gnome-terminal/stable), and like xterm that does work, no issue for that terminal.

So, iTerm2 specific issue?

Other exotic terminals may be interesting to see: kitty, alacritty.

**Comment by bfredl** [comment]
What `$TERM` does item2 use? Chances are that iterm2 doesn't do bold-bright by itself, so that git changes the codes used for `%C(bold green)` to do this with this terminfo. Alternatively the iterm2 terminfo causes nvim TUI to do something funky, but I don't have a mac so I can't debug it.

**Comment by clason** [comment]
@bfredl iTerm2 uses
```
TERM xterm-256color
TERM_PROGRAM iTerm.app
```

see also https://gitlab.com/gnachman/iterm2/issues/7854

**Comment by justinmk** [comment]
Does `TERM=iterm2` make any difference?

Did something regress since #10994, or this is just a report that it's not perfect for iTerm2?

**Comment by clason** [comment]
> Does `TERM=iterm2` make any difference?

That breaks the shell, and it does not.

> Did something regress since #10994, or this is just a report that it's not perfect for iTerm2?

The latter (at least for me).

**Comment by justinmk** [comment]
> however quite a few utility which use **bolded named colors** are broken.

"Broken" meaning they show soft colors instead of bright/bold? People sure throw the word "broken" around a lot these days.

This sounds like an iterm2 issue, closing until proven otherwise. Patch welcome also.

**Comment by bluz71** [comment]
Doing `git log --format='%C(bold green)%h%C(bold red)%d%C(reset)'` in a standard iTerm2 terminal emits correct colors.

Doing `git log --format='%C(bold green)%h%C(bold red)%d%C(reset)'` inside a Neovim terminal within iTerm2, with `termguicolors` set, after #10994 was merged, produces incorrect colors.

> "Broken" meaning they show soft colors instead of bright/bold?

Bold Green will be displayed as Green (color 10 will be displayed as color 2). If broken was too strong a term, I apologize. Replace all instances of *broken* with *incorrect*.

What is required to re-open this? Myself and @clason are here to collaborate and improve Neovim, this issue included.

If iTerm2 is the issue, and maybe it is, then why does coloured Git log work in the normal terminal but not within the terminal within Neovim that inherits colors (PR #10994)? That more likely seems a Neovim + libvterm + termguicolors + #10994 issue does it not?

I should also note that #7018 was closed as part of the 0.4.0 release but I can assure you that it was not corrected. It is still an issue that is partly solved by #10994. One of the parts to solve is this issue.

I believe this should be re-opened until it is proven it is an iTerm2 issue. When that is proven I will open an issue at the iTerm2 issue tracker.

**Comment by bfredl** [comment]
> What is required to re-open this? Myself and @clason are here to collaborate and improve Neovim, this issue included.

Anyone can open/suggest a PR at anytime even if the issue is closed (or if there is no issue about something at all).

> If iTerm2 is the issue, and maybe it is, then why does coloured Git log work in the normal terminal but not within the terminal within Neovim that inherits colors (PR #10994)? That more likely seems a Neovim + libvterm + termguicolors + #10994 issue does it not?

Nvim is supposed to pass colors right through. Otherwise could be proven by showing some color sequence when input to nvim, gives the wrong output back to the host terminal.

> I believe this should be re-opened until it is proven it is an iTerm2 issue. When that is proven I will open an issue at the iTerm2 issue tracker

The "gold standard" for `TERM=xterm-whatever` terminals is xterm itself, and it works there. If iterm2 has chosen to use `TERM=xterm` but behaves differently it is a iterm2 issue pretty much by definition (but if other terminals have the same deviance, it a more prioritized issue of course).

**Comment by justinmk** [comment]
Doesn't it work with other terminals? It may be a terminfo issue. Moreoever,

    git log --format='%C(bold green)%h%C(bold red)%d%C(reset)'

looks the same for me in iTerm2 inside/outside of Nvim with `set tgc` + `:terminal`. Note: 

- `TERM=xterm-256color`
- iTerm2 version 3.2.9

**Comment by clason** [comment]
(I also apologize for using "breaking" too glibly; fish outputs a bunch of warnings but of course still performs its main functions correctly.)

The issues occur with `xterm-256color` as well; this is on 3.3.7beta4 (the latest). It might depend on the color scheme used in the profile (a screenshot of the colors tab of the used profile may be useful, as well as testing if a clean install of iTerm helps as in the linked issue).

I should also mention that the same issue occurs in VimR, although that's still based on nvim 0.4.2, which doesn't have #10994.

**Comment by bluz71** [comment]
I will do my best to install and test both Kitty and alacritty on my Macbook.

@justinmk, do you different colors set for Green (2) and Bold Green (10)?
I set my own custom terminal colors for 1-16.
Hmmm, I also set my own Neovim theme colors 1-16 also. I may need to explore that also.

There are enough avenues to explore here, so I have some homework to do.

If I can repeat this issue in other terminals on Mac then I have a stronger case to re-open this.
If I can provide a test case for iTerm2 I will do so.

I am busy today, but I will report back soon.

Thanks.

**Comment by bluz71** [comment]
@clason,

Theme could be a factor, maybe maybe not.
I use my own theme AND I set Neovim terminal colors 1 - 16. I will explore this for sure.

I feel we are very close to solving #7018, very close.

Note, I am using `TERM=xterm-256color` on iTerm2.

**Comment by justinmk** [comment]
>  do you different colors set for Green (2) and Bold Green (10)?

with `nvim +"set tgc"` and `:terminal`, and iTerm2 bold green set to light pink:

![Screen Shot 2019-11-10 at 15 34 30](https://user-images.githubusercontent.com/1359421/68552773-b8b1e000-03cf-11ea-9062-6119c6b3cbfc.png)

**Comment by bluz71** [comment]
Interesting. That appears to work.
Maybe the Vim theme plays a part?
I will test.

**Comment by clason** [comment]
@bluz71 Sure, the colorscheme can set its own terminal colors (in which case iTerm's are ignored), and disable bold.

**Comment by bluz71** [comment]
Ok guys,

The theme **does** make a difference. That is the driving issue here.
I use my own [moonfly](https://github.com/bluz71/vim-moonfly-colors) theme which explicity sets Neovim terminal colors as noted here:

```viml
let g:terminal_color_0  = "#373c40"
let g:terminal_color_1  = "#ff5454"
let g:terminal_color_2  = "#8cc85f"
let g:terminal_color_3  = "#e3c78a"
let g:terminal_color_4  = "#80a0ff"
let g:terminal_color_5  = "#ce76e8"
let g:terminal_color_6  = "#7ee0ce"
let g:terminal_color_7  = "#de935f"
let g:terminal_color_8  = "#f09479"
let g:terminal_color_9  = "#f74782"
let g:terminal_color_10 = "#42cf89"
let g:terminal_color_11 = "#cfcfb0"
let g:terminal_color_12 = "#78c2ff"
let g:terminal_color_13 = "#ae81ff"
let g:terminal_color_14 = "#85dc85"
let g:terminal_color_15 = "#e2637f"
```
Those same colors are also set at the terminal level.

Those settings directly have an effect on this issue.
Once I comment them out, all good, no issue, similar to what @justinmk notes above.
However, if I set just these two colors at the Neovim command line:

```viml
:let g:terminal_color_2  = "#8cc85f"
:let g:terminal_color_10 = "#42cf89"
```

And then do `git log --format='%C(bold green)%h%C(bold red)%d%C(reset)'` in a Neovim terminal I see the terminal_2 color (green) being emitted instead of terminal_10 (bold/bright green).

Ok, we have an Neovim-specific internal interaction issue.

With proper terminal inheritance, as implemented by @bfredl, I tend to believe the `g:terminal_color` settings are possibly more trouble than benefit?

Should I remove them completely?
Keep them?
Keep them only if `:termguicolors` is **not** set?
Keep them only for Neovim versions prior to 0.5.0?
Or maybe Neovim 0.5.0 and later should ignore them?
Is the recent terminal inheritance work just for `:termguicolors`? or for both `cterm` and `gui`?

This is not an iTerm2 bug, hence I will not install Kitty or Alacritty.

**Comment by bluz71** [comment]
Ok, tested on 0.4.3 and it also uses the wrong green for `git log --format='%C(bold green)%h%C(bold red)%d%C(reset)'` when `g:terminal_colors` are set within `:terminal` under iTerm2.

So I don't believe this is a new introduced issue.
It's an older issue.

@bfredl term inheritance is working, once I comment out `g:terminal_colors`.

`g:terminal_colors` appears to have an issue with bold colors on iTerm2.
Under Gnome Terminal there is **no** issue with `g:terminal_colors`.

Could be an iTerm2 bug or a `g:terminal_colors`-on-iTerm2 bug.

`TERM=xterm-256color`.

**Comment by bluz71** [comment]
Kitty terminal on Mac also displays the wrong colors, much like iTerm2.

Steps to reproduce the issue:
- Open Kitty or iTerm2
- Launch nvim 0.4.3
- set `termguicolors`
- `:let g:terminal_color_2  = "#8cc85f"` (green)
- `:let g:terminal_color_10 = "#42cf89"` (bold green)
- Open a terminal, `:terminal`
- Enter `git log --format='%C(bold green)%h%C(bold red)%d%C(reset)'`
- Notice that *green* (color 2) will be emitted for the Git hashes instead of *bold green* (color 10).

Kitty version: 0.14.6, `TERM=xterm-kitty`
iTerm2 version: 3.3.6, `TERM=xterm-256color`

**Comment by bfredl** [comment]
> The theme does make a difference. That is the driving issue here.

This why the issue template requires reproductions with `nvim -u NORC`. With unknown user config anything can happen.

> Should I remove them completely?
Keep them?
Keep them only if :termguicolors is not set?
Keep them only for Neovim versions prior to 0.5.0?
Or maybe Neovim 0.5.0 and later should ignore them?

This is up to you. The options are there for those who want to change the terminal colors, and it is the _only_ way if you use GUI with `:terminal`, so they should stay in Neovim.

**Comment by clason** [comment]
So let me get this straight: The issue is that with `g:terminal_color`s set, the terminal does not use bright/bold colors but falls back to the non-bold ones? That does sound like a bug to me. 

I can reproduce that in Konsole `xterm-256colors` (bright colors used in terminal, non-bright in vim :terminal if and only if `g:terminal_color` is set). 

> This why the issue template requires reproductions with nvim -u NORC. With unknown user config anything can happen.

:100: Would have saved me some time chasing the wrong squirrel, too.

**Comment by bfredl** [comment]
> So let me get this straight: The issue is that with g:terminal_colors set, the terminal does not use bright/bold colors but falls back to the non-bold ones? That does sound like a bug to me.

There is no such thing as a "bold" color. Some terminals convert bold attr to bright colors, some do not. The only way nvim can support both is by forwarding the colors, which is does by default since 0.5. Setting  `g:terminal_colors` is literally asking nvim to _not_ behave like the host terminal, how is it a bug that nvim doesn't reflect the host terminal when you just asked it not do?

**Comment by clason** [comment]
I apologize, from the OP's description I had assumed that `%C(bold green)` was mapped to `terminal_color_10`. If that is wrong, it is of course not a bug. (Teaches me to a) assume things and b) stick my nose where it doesn't belong. I'll shut up now.)

**Comment by bluz71** [comment]
The discussion in this thread has proven very useful.

In my own moonfly Vim theme I  now skip the setting of `g:terminal_colors` but only if TUI Neovim 0.5.0, or later, is detected with `termguicolors` set. That version contains pass terminal palette pass-through which works great.

I do that check as follows:

```viml
if has("nvim-0.5.0") && &termguicolors && nvim_list_uis()[0]['ext_termcolors']
  " Do NOT set g:terminal_colors
else
  " Do set g:terminal_colors
endif
```

Detecting TUI vs GUI Neovim is hard, @justinmk in another issue suggested the `nvim_list` check above. It appears to work with terminal Neovim vs goneovim.

GUIs don't inherit terminal colors, hence the g:terminal_colors settings are still needed even for GUIs that build upon Neovim 0.5.0. There is no terminal to inherit from, there is no passing through.

Yes, there is no issue to fix here.
Sorry for taking all your time, but in the end the clarity provided here is useful especially for a theme authors.

Thanks.

**Comment by liushapku** [comment]
I have exactly the same problem. I am using WSL2 in windows terminal, with nvim 0.43. 
`git log --format='%C(bold green)%h%C(bold red)%d%C(reset)'` shows `(green)` and `(red)`. 

In my `.vimrc`, I have set the `termguicolors` and also set `g:terminal_color` explicitly to match my terminal palette. 

If I remove these settings, `(bold green)` is still displayed as `(green)`. But the palette is not inherited from the terminal any more

**Comment by bluz71** [comment]
I recommend using RGB colours for your Git log statement. It is what I ended up doing t solve this issue, for example:

```
git log --graph --date=human --format='%C(#e3c78a)%h%C(#ff5454)%d%C(reset) - %C(#36c692)(%ad)%C(reset) %s %C(#80a0ff){%an}%C(reset)'" 
```

I do the same for fzf.

**Comment by xinkeleliu** [comment]
> I recommend using RGB colours for your Git log statement. It is what I ended up doing t solve this issue, for example:
> 
> ```
> git log --graph --date=human --format='%C(#e3c78a)%h%C(#ff5454)%d%C(reset) - %C(#36c692)(%ad)%C(reset) %s %C(#80a0ff){%an}%C(reset)'" 
> ```
> 
> I do the same for fzf.

`git` is only an example. Actually, all programs that uses color are affected, like `ls`, `tree`, `ag`, .... 
The problem can also be demonstrated using `LS_COLORS="di:1;31" ls` vs `LS_COLORS="di:31" ls` and any other program that uses `LS_COLORS`
